design/httpproxy-route-sorting.md

```diff
@@ -0,0 +1,186 @@
+
+* Cluster Administrator: A cluster administrator is responsible for providing core functionality to a k8s cluster. For the purposes of this document the cluster administrator installs Contour to the cluster. The administrator is responsible for the health of the Contour services.
+
+* ResourceOwner: Someone who manages an HTTPProxy resource inside the Kubernetes cluster configuring the Route/Service combination to allow users of the application to access. A resource owner might not be familiar with Envoy, doesn't have access to the Envoy admin interface and has surface level understanding of how the the HTPProxy Routes are enforced.
+
+## Goals
+
+- Allow `Cluster Administrator` to disable the sorting logic
+
+## Non-goals
+Improving or changing the heuristics used for sorting routes.
+
+## High-level design
+
+At a high level we want to offer a configuration for `Cluster Administrator` to be abe to disable route sorting. From the configurator perspective they would have access to a `featureFlag` called `omitRouteSorting` that disables all route sorting. This can be implemented similar to the feature gate we used for [Endpoint Slices](https://github.com/projectcontour/contour/pull/5745)
+
+For Contour developers this means that we need to be careful of the data-structures we use to store the routes as we translate them from `HTTPProxy` to Envoy `Routes`. With sorting enabled we could use data structures like a `map` that don't offer any ordering guarantees but this wouldn't matter since Contour decided the order. To avoid this we need to switch to using `slices` as the only (Golang native) data structure for all cases. This might make some operations theoretically inefficient but in practice the number of possible routes within a virtual host is pretty small so it shouldn't matter. This invariance can be guaranteed with unit tests which already cover this.
+
+### Basic Case
+
+In the basic case a single `HTTPProxy` with the following routes:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example
+spec:
+  virtualhost:
+    fqdn: example.com
+  routes:
+    - conditions:
+      - prefix: /
+      - header:
+          exact: bar
+          name: x-experiment
+      services:
+        - name: b1
+          port: 400
+    - conditions:
+      - exact: /bar
+      services:
+        - name: b2
+          port: 400
+    - conditions:
+      - exact: /
+      services:
+        - name: b3
+          port: 400
+```
+
+Would result in the following:
+
+| Request                          | With Sorting  | Without Sorting |
+| -------------                    | ------------- | --------------- |
+| GET -H "x-experiment: bar" /bar  | b2            | b1              |
+| GET /bar | Content Cell          | b2            | b2              |
+| GET /                            | b3            | b3              |
+
+
+### Route Matching with Include Statements
+
+Include statements are resolved in a depth first fashion.
```

Comment:
Here the "include" routes are sorted before the routes directly defined in the root HTTPProxy; any particular justification for that order vs. putting them after the directly-defined routes?

Reply:
> Here the "include" routes are sorted before the routes directly defined in the root HTTPProxy; any particular justification for that order vs. putting them after the directly-defined routes?

I dont 100% follow, does `any particular justification for that order` refer to the example below or the `depth first` resolution?

Reply:
Sorry, put the comment on the wrong line. It's about the example below, where the route(s) defined by `includes` come before all the routes defined directly on the proxy in `routes` (i.e. `/blog` comes first in the Envoy route table, followed by `/bar` and `/`).

The depth-first bit for included routes makes sense to me.